### PR TITLE
StatefulSets rolling upgrades

### PR DESCRIPTION
**Pull request by sebgl**
Handle cluster rolling upgrades with StatefulSets.

Relies on manually updating the `rollingUpdate.Partition` index of a sset when we're ready to restart the ES node.
For now, it only handles 1 node restart at a time, mostly because we rely on cluster health green checks.
It also introduces 2 additional concepts:

* StatefulSet updates expectations, based on UID + Generation: there are certain ES calls we don't want to make if our StatefulSet cache is out of date
* "lazy" ES client calls: some ES calls we do to handle the upgrade cannot rely on out-of-date information from observers. We need fresh ES information on the state of the cluster before eg. re-enabling shard allocations. However we don't want to perform these calls at every single reconciliation. I see this as a "stop-gap" implementation before we go on with a larger refactoring of sync/async ES calls.

This is still WIP since many things are still missing:
* changeBudget considerations.
* zen1/zen2 handling.
* Unit tests, everywhere. I still think a large portion of this code may be refactored in the short-term to account for change budget & zen1/zen2. Keeping unit tests for a later PR.
* E2E tests, kept for a later PR.

### CODE REVIEW

File: operators/pkg/controller/elasticsearch/driver/default.go [hunk 1]
```diff
@@ -327,7 +274,7 @@ func (d *defaultDriver) Reconcile(
 		)
 	}
 
-	res = d.reconcileNodeSpecs(es, podTemplateSpecBuilder, esClient, observedState)
+	res = d.reconcileNodeSpecs(es, esReachable, podTemplateSpecBuilder, esClient, observedState)
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/driver/default.go]: Is there a particular reason to not check esReachable and requeue the reconcilliation before calling `reconcileNodeSpecs`?

File: operators/pkg/controller/elasticsearch/client/v6.go [hunk 2]
```diff
@@ -29,18 +34,50 @@ func (c *clientV6) UpdateSettings(ctx context.Context, settings Settings) error
...
+			Cluster: ClusterRoutingSettings{
+				Routing: RoutingSettings{
+					Allocation: RoutingAllocationSettings{
+						Enable: "primaries",
+					},
+				},
+			},
+		},
+	}
+	return c.put(ctx, "/_cluster/settings", allocationSettings, nil)
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/client/v6.go]: Code is factorizable between `EnableShardAllocation` and `DisableReplicaShardsAllocation`.

File: operators/pkg/controller/elasticsearch/driver/default.go (hunk 1 above)
Comment (sebgl) [reply operators/pkg/controller/elasticsearch/driver/default.go]: We still want to create additional pods we can create: this does not require the cluster to respond to requests.

2 main use cases:
1. There are 0 nodes in the cluster, we need to create the first ones :)
2. User requests cluster scale-up, we can do it even though ES does not respond.

File: operators/pkg/controller/elasticsearch/driver/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+	// Make sure all pods scheduled for upgrade have been upgraded.
+	scheduledUpgradesDone, err := sset.ScheduledUpgradesDone(d.Client, statefulSets)
+	if err != nil {
+		return results.WithError(err)
+	}
+	if !scheduledUpgradesDone {
+		log.V(1).Info(
+			"Rolling upgrade not over yet, some pods don't have the updated revision, keeping shard allocations disabled",
+			"namespace", es.Namespace,
+			"name", es.Name,
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/driver/upgrade.go]: In #1168, we chose to be explicit with the owner of the name and log `"es_name", es.Name`.

File: operators/pkg/controller/elasticsearch/driver/default.go (hunk 1 above)
Comment (thbkrkr) [reply operators/pkg/controller/elasticsearch/driver/default.go]: :ok_hand: :no_mouth:

File: operators/pkg/controller/elasticsearch/driver/esstate.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
+package driver
```
Comment (pebrc) [inline operators/pkg/controller/elasticsearch/driver/esstate.go]: license header should be on top?

File: operators/pkg/controller/elasticsearch/client/v6.go (hunk 2 above)
Comment (sebgl) [reply operators/pkg/controller/elasticsearch/client/v6.go]: Yes! I'll merge that one and fix it in https://github.com/elastic/cloud-on-k8s/pull/1262.

File: operators/pkg/controller/elasticsearch/driver/upgrade.go (hunk 3 above)
Comment (sebgl) [reply operators/pkg/controller/elasticsearch/driver/upgrade.go]: Yes! I'll merge that one and fix it in #1262.

File: operators/pkg/controller/elasticsearch/driver/esstate.go (hunk 4 above)
Comment (sebgl) [reply operators/pkg/controller/elasticsearch/driver/esstate.go]: Yes! I'll merge that one and fix it in #1262.

**Pull request merged**